python/cudf/cudf/core/resample.py

```diff
@@ -0,0 +1,360 @@
+            )
+
+        # get the start and end values that will be used to generate
+        # the bin labels
+        min_date, max_date = key_column._minmax()
+        start, end = _get_timestamp_range_edges(
+            pd.Timestamp(min_date.value),
+            pd.Timestamp(max_date.value),
+            offset,
+            closed=closed,
+        )
+
+        # in some cases, an extra time stamp is required in order to
+        # bin all the values. It's OK if we generate more labels than
+        # we need, as we remove any unused labels below
+        end += offset
+
+        # generate the labels for binning the key column:
+        bin_labels = cudf.date_range(start=start, end=end, freq=freq,)
+
+        # We want the (resampled) column of timestamps in the result
+        # to have a resolution closest to the resampling
+        # frequency. For example, if resampling from '1T' to '1s', we
+        # want the resulting timestamp column to by of dtype
+        # 'datetime64[s]'.  libcudf requires the bin labels and key
+        # column to have the same dtype, so we compute a `result_type`
+        # and cast them both to that type.
+        try:
+            from cudf.core.tools.datetimes import (
+                _offset_alias_to_code,
+                _unit_dtype_map,
+            )
+
+            result_type = np.dtype(
+                _unit_dtype_map[_offset_alias_to_code[offset.name]]
+            )
+        except KeyError:
+            # unsupported resolution (we don't support resolutions >s)
+            # fall back to using datetime64[s]
+            result_type = np.dtype("datetime64[s]")
+
+        # TODO: Ideally, we can avoid one cast by having `date_range`
+        # generate timestamps of a given dtype.  Currently, it can
+        # only generate timestamps with 'ns' precision
+        key_column = key_column.astype(result_type)
+        bin_labels = bin_labels.astype(result_type)
+
+        # bin the key column:
+        bin_numbers = cudf._lib.labeling.label_bins(
+            key_column,
+            left_edges=bin_labels[:-1]._column,
+            left_inclusive=(closed == "left"),
+            right_edges=bin_labels[1:]._column,
+            right_inclusive=(closed == "right"),
+        )
+
+        if label == "right":
+            bin_labels = bin_labels[1:]
+        else:
+            bin_labels = bin_labels[:-1]
+
+        # if we have more labels than bins, remove the extras labels:
+        nbins = bin_numbers.max() + 1
+        if len(bin_labels) > nbins:
+            bin_labels = bin_labels[:nbins]
+
+        bin_labels.name = self.names[0]
+        self.bin_labels = bin_labels
+
+        # replace self._key_columns with the binned key column:
+        self._key_columns = [
+            bin_labels._gather(bin_numbers, check_bounds=False)._column.astype(
+                result_type
+            )
+        ]
+
+
+# NOTE: this function is vendored from Pandas
+def _get_timestamp_range_edges(
```

Comment:
@brandon-b-miller @isVoid the last time I spoke with @shwina I mentioned that I might actually prefer to use pandas internal APIs rather than vendor them. Here's how I see the two options:
1. We vendor the functions. The code will always work, even if pandas changes the signature of these functions. However, if pandas changes the behavior of these functions without changing the signature, our behavior could silently diverge from pandas unless our test coverage is good enough to catch all edge cases (which is basically impossible given how many edge cases there are in pandas APIs).
2. We use pandas internal APIs. If pandas changes the signature of these functions, cudf immediately fails, but it fails loudly. Meanwhile, we are automatically kept in sync with pandas behavior changes.

It seems to me that if our goal is to match all behavior of pandas, we should prefer option (2) even if it means using pandas internals since that's effectively what we're doing anyway. Thoughts?

Reply:
I think it ends up being conservation of misery at some point because even if we import the functions so that we can keep up to date with any changes, they could be fully moved somewhere else at some point or even removed completely. So we trade the maintenance burden of keeping them up to date ourselves with the risk of having to go find them or even vendor them later anyways if they disappear and aren't replaced with something suitable for our uses. 

On the other hand vendoring them creates a "tribal knowledge" situation that makes it harder for contributors to know how to maintain the codebase along with the divergence issue you mentioned. Maybe we can judge it on a case by case basis? In the case of `_get_timestamp_range_edges` it looks like the code has changed as little as 6 months ago but the function has lived in the file for almost a decade, based on the blame. I suppose given that I might lean towards importing it...

Reply:
Vendoring also lets us tweak/optimize things for cuDF. For example, in these vendored functions, we can safely remove much of the timezone-related logic and not lose any cuDF functionality.

Reply:
sync'd offline and we decided to leave these vendored for the moment, recognizing that there's no good solution.